In [0]:
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
import pandas as pd

In [0]:
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        do_value=0.02
        self.conv1 = nn.Sequential(
            nn.Conv2d(in_channels=1, out_channels=20, kernel_size=(3,3), padding=0, stride=1), 
            nn.BatchNorm2d(20),
            nn.Dropout(do_value),
            nn.ReLU()
        )
        #Output=26 RF=3X3 [RFin + (Ksize-1 * JMPin) => 1+(3-1)*1 =3]  :JMPin=1, Jout= JMPin X s = 1
        
        self.conv2 = nn.Sequential(
            nn.Conv2d(in_channels=20, out_channels=16, kernel_size=(3,3), padding=0, stride=1),
            nn.BatchNorm2d(16),
            nn.Dropout(do_value),
            nn.ReLU()    
        )
        #Output=24 RF=5X5  [RFin + (Ksize-1 * JMPin) => 3+(3-1)*1 =5] :JMPin=1, Jout =JMPin X s =1

        self.pool1 = nn.MaxPool2d(2, 2)#Output=12 RF=6X6 [RFin + (Ksize-1 * JMPin) => 5+(2-1)*1 =6] :JMPin=1, Jout=  JMPin X s =2

        self.conv3 = nn.Sequential(
            nn.Conv2d(in_channels=16, out_channels=16, kernel_size=(3,3), padding=0, stride=1),
            nn.BatchNorm2d(16),
            nn.Dropout(do_value),
            nn.ReLU()
        )
        #Output=10 RF=10X10 [RFin + (Ksize-1 * JMPin) => 6+(3-1)*2 =10] : Jout= JMPin X s = 2X1 :JMPin=2, Jout= JMPin X s = 2X1=2

        self.pool2 = nn.MaxPool2d(2, 2) #Output=5 RF=12[RFin + (Ksize-1 * JMPin) => 10+(2-1)*2 =12]  :JMPin=2, Jout =JMPin X s = 2X2 =4

        self.conv4 = nn.Sequential(
            nn.Conv2d(in_channels=16, out_channels=16, kernel_size=(3,3), padding=0, stride=1),
            nn.BatchNorm2d(16),
            nn.Dropout(do_value),
            nn.ReLU()            
        )
        #Output=3 RF= 20[RFin + (Ksize-1 * JMPin) => 12+(3-1)*4 =20] :JMPin=4, Jout =JMPin X s = 4X1=4
        
        self.conv5 = nn.Sequential(
            nn.Conv2d(in_channels=16, out_channels=13, kernel_size=(3,3), padding=0, stride=1),
            nn.BatchNorm2d(13),
            nn.Dropout(do_value),
            nn.ReLU()
            
        )
        #Output=1 RF=28 [RFin + (Ksize-1 * JMPin) => 20+(3-1)*4 =28]  :JMPin=4, Jout=JMPin X s = 4X1=4
        
        self.conv6 = nn.Sequential(
            nn.Conv2d(in_channels=13, out_channels=10, kernel_size=(1,1), padding=0, stride=1)
            
        )
        
     

    def forward(self, x):
        x = self.conv1(x)
        x = self.conv2(x)
        x = self.pool1(x)
        x = self.conv3(x)
        x = self.pool2(x)
        x = self.conv4(x)
        x = self.conv5(x)   
        x = self.conv6(x)      
        x = x.view(-1, 10)
        return F.log_softmax(x)

In [5]:
!pip install torchsummary
from torchsummary import summary
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")
model = Net().to(device)
summary(model, input_size=(1, 28, 28))

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1           [-1, 20, 26, 26]             200
       BatchNorm2d-2           [-1, 20, 26, 26]              40
           Dropout-3           [-1, 20, 26, 26]               0
              ReLU-4           [-1, 20, 26, 26]               0
            Conv2d-5           [-1, 16, 24, 24]           2,896
       BatchNorm2d-6           [-1, 16, 24, 24]              32
           Dropout-7           [-1, 16, 24, 24]               0
              ReLU-8           [-1, 16, 24, 24]               0
         MaxPool2d-9           [-1, 16, 12, 12]               0
           Conv2d-10           [-1, 16, 10, 10]           2,320
      BatchNorm2d-11           [-1, 16, 10, 10]              32
          Dropout-12           [-1, 16, 10, 10]               0
             ReLU-13           [-1, 16, 10, 10]               0
        MaxPool2d-14             [-1, 1

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:67: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.


In [0]:


torch.manual_seed(1)
batch_size = 64

kwargs = {'num_workers': 1, 'pin_memory': True} if use_cuda else {}
train_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=True, download=True,
                    transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ])),
    batch_size=batch_size, shuffle=True, **kwargs)
test_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=False, transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ])),
    batch_size=batch_size, shuffle=True, **kwargs)


In [0]:
from tqdm import tqdm

Train_Loss=[]
Train_Accuracy=[]
Test_Loss=[]
Test_Accuracy=[]

def train(model, device, train_loader, optimizer, epoch):
    model.train()
    pbar = tqdm(train_loader)
    correct_train=0
    for batch_idx, (data, target) in enumerate(pbar):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        Train_Loss.append(loss.item())
        pred = output.argmax(dim=1, keepdim=True)
        correct_train += pred.eq(target.view_as(pred)).sum().item()
        Train_Accuracy.append(100.00 *correct_train/len(train_loader.dataset))
        loss.backward()
        optimizer.step()
        #pbar.set_description(desc= f'TRAINING Loss={loss.item()} batch_id={batch_idx}')
        
        ##Added detailed percentage:
        pbar.set_description(desc= f'TRAIN Loss={loss.item()} batch_id={batch_idx} Correct={correct_train} / {len(train_loader.dataset)} TRAIN ACCURACY={100.00 *correct_train/len(train_loader.dataset)}')
     


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            loss=F.nll_loss(output, target)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            
            correct += pred.eq(target.view_as(pred)).sum().item()
            
    test_loss /= len(test_loader.dataset)
    Test_Loss.append(test_loss)
    Test_Accuracy.append(100. * correct / len(test_loader.dataset))
    print('\nTEST: Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))
    


In [8]:
from torch.optim.lr_scheduler import StepLR

model = Net().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
scheduler = StepLR(optimizer, step_size=6, gamma=0.1)

df_Test_Train_Analysis=pd.DataFrame(columns=['EPOCH','Test_Acc','Train_Acc','Acc_Diff'])
print(df_Test_Train_Analysis.shape)
for epoch in range(1, 15):
    print('EPOCH #',epoch)
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)
    print('----------------------------------------------------------------------')

  0%|          | 0/938 [00:00<?, ?it/s]

(0, 4)
EPOCH # 1


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:67: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.
TRAIN Loss=0.11304358392953873 batch_id=937 Correct=55830 / 60000 TRAIN ACCURACY=93.05: 100%|██████████| 938/938 [00:15<00:00, 59.92it/s]
  0%|          | 0/938 [00:00<?, ?it/s]


TEST: Average loss: 0.0468, Accuracy: 9862/10000 (98.62%)

----------------------------------------------------------------------
EPOCH # 2


TRAIN Loss=0.07809174060821533 batch_id=937 Correct=58795 / 60000 TRAIN ACCURACY=97.99166666666666: 100%|██████████| 938/938 [00:15<00:00, 61.02it/s]
  0%|          | 0/938 [00:00<?, ?it/s]


TEST: Average loss: 0.0355, Accuracy: 9902/10000 (99.02%)

----------------------------------------------------------------------
EPOCH # 3


TRAIN Loss=0.044644542038440704 batch_id=937 Correct=59060 / 60000 TRAIN ACCURACY=98.43333333333334: 100%|██████████| 938/938 [00:15<00:00, 61.15it/s]
  0%|          | 0/938 [00:00<?, ?it/s]


TEST: Average loss: 0.0362, Accuracy: 9904/10000 (99.04%)

----------------------------------------------------------------------
EPOCH # 4


TRAIN Loss=0.03305545449256897 batch_id=937 Correct=59225 / 60000 TRAIN ACCURACY=98.70833333333333: 100%|██████████| 938/938 [00:15<00:00, 60.10it/s]
  0%|          | 0/938 [00:00<?, ?it/s]


TEST: Average loss: 0.0296, Accuracy: 9911/10000 (99.11%)

----------------------------------------------------------------------
EPOCH # 5


TRAIN Loss=0.002207130193710327 batch_id=937 Correct=59277 / 60000 TRAIN ACCURACY=98.795: 100%|██████████| 938/938 [00:15<00:00, 59.94it/s]
  0%|          | 0/938 [00:00<?, ?it/s]


TEST: Average loss: 0.0281, Accuracy: 9908/10000 (99.08%)

----------------------------------------------------------------------
EPOCH # 6


TRAIN Loss=0.004798993468284607 batch_id=937 Correct=59370 / 60000 TRAIN ACCURACY=98.95: 100%|██████████| 938/938 [00:15<00:00, 59.74it/s]
  0%|          | 0/938 [00:00<?, ?it/s]


TEST: Average loss: 0.0261, Accuracy: 9926/10000 (99.26%)

----------------------------------------------------------------------
EPOCH # 7


TRAIN Loss=0.01608934998512268 batch_id=937 Correct=59393 / 60000 TRAIN ACCURACY=98.98833333333333: 100%|██████████| 938/938 [00:15<00:00, 60.06it/s]
  0%|          | 0/938 [00:00<?, ?it/s]


TEST: Average loss: 0.0242, Accuracy: 9930/10000 (99.30%)

----------------------------------------------------------------------
EPOCH # 8


TRAIN Loss=0.1274120956659317 batch_id=937 Correct=59445 / 60000 TRAIN ACCURACY=99.075: 100%|██████████| 938/938 [00:15<00:00, 59.67it/s]
  0%|          | 0/938 [00:00<?, ?it/s]


TEST: Average loss: 0.0223, Accuracy: 9937/10000 (99.37%)

----------------------------------------------------------------------
EPOCH # 9


TRAIN Loss=0.0036022067070007324 batch_id=937 Correct=59499 / 60000 TRAIN ACCURACY=99.165: 100%|██████████| 938/938 [00:15<00:00, 60.31it/s]
  0%|          | 0/938 [00:00<?, ?it/s]


TEST: Average loss: 0.0232, Accuracy: 9931/10000 (99.31%)

----------------------------------------------------------------------
EPOCH # 10


TRAIN Loss=0.0008698999881744385 batch_id=937 Correct=59530 / 60000 TRAIN ACCURACY=99.21666666666667: 100%|██████████| 938/938 [00:15<00:00, 58.81it/s]
  0%|          | 0/938 [00:00<?, ?it/s]


TEST: Average loss: 0.0235, Accuracy: 9938/10000 (99.38%)

----------------------------------------------------------------------
EPOCH # 11


TRAIN Loss=0.01777687668800354 batch_id=937 Correct=59532 / 60000 TRAIN ACCURACY=99.22: 100%|██████████| 938/938 [00:15<00:00, 59.56it/s]
  0%|          | 0/938 [00:00<?, ?it/s]


TEST: Average loss: 0.0250, Accuracy: 9931/10000 (99.31%)

----------------------------------------------------------------------
EPOCH # 12


TRAIN Loss=0.04591198265552521 batch_id=937 Correct=59578 / 60000 TRAIN ACCURACY=99.29666666666667: 100%|██████████| 938/938 [00:15<00:00, 60.17it/s]
  0%|          | 0/938 [00:00<?, ?it/s]


TEST: Average loss: 0.0257, Accuracy: 9925/10000 (99.25%)

----------------------------------------------------------------------
EPOCH # 13


TRAIN Loss=0.00904218852519989 batch_id=937 Correct=59592 / 60000 TRAIN ACCURACY=99.32: 100%|██████████| 938/938 [00:15<00:00, 60.24it/s]
  0%|          | 0/938 [00:00<?, ?it/s]


TEST: Average loss: 0.0208, Accuracy: 9936/10000 (99.36%)

----------------------------------------------------------------------
EPOCH # 14


TRAIN Loss=0.008609741926193237 batch_id=937 Correct=59600 / 60000 TRAIN ACCURACY=99.33333333333333: 100%|██████████| 938/938 [00:15<00:00, 59.99it/s]



TEST: Average loss: 0.0229, Accuracy: 9940/10000 (99.40%)

----------------------------------------------------------------------


#Goal	
In this iteration we intend to increase Model Accuracy to 99.4% and beyond at the same time maintaining the model consistancy.
#Params	
9,923

#WITH 15 EPOCHS

#DropOut value
0.02

#Best Train Accuracy	
99.33%

#Best Test Accuracy	
99.4%

#Observation/ Analysis/Conclusion	
We have achived target accuracy also the Model did not fall to Ovefitting.
This is a decent model.


